pageserver/src/tenant/timeline.rs

```diff
@@ -3081,34 +3194,55 @@ impl Timeline {
                             key = key.next();
                             continue;
                         }
-                        let img = match self.get(key, lsn, ctx).await {
-                            Ok(img) => img,
-                            Err(err) => {
-                                // If we fail to reconstruct a VM or FSM page, we can zero the
-                                // page without losing any actual user data. That seems better
-                                // than failing repeatedly and getting stuck.
-                                //
-                                // We had a bug at one point, where we truncated the FSM and VM
-                                // in the pageserver, but the Postgres didn't know about that
-                                // and continued to generate incremental WAL records for pages
-                                // that didn't exist in the pageserver. Trying to replay those
-                                // WAL records failed to find the previous image of the page.
-                                // This special case allows us to recover from that situation.
-                                // See https://github.com/neondatabase/neon/issues/2601.
-                                //
-                                // Unfortunately we cannot do this for the main fork, or for
-                                // any metadata keys, keys, as that would lead to actual data
-                                // loss.
-                                if is_rel_fsm_block_key(key) || is_rel_vm_block_key(key) {
-                                    warn!("could not reconstruct FSM or VM key {key}, filling with zeros: {err:?}");
-                                    ZERO_PAGE.clone()
-                                } else {
-                                    return Err(err);
-                                }
+
+                        key_request_accum.add_key(key);
+                        if key_request_accum.size() >= Timeline::MAX_GET_VECTORED_KEYS
+                            || key.next() == range.end
+                        {
+                            let results = self
+                                .get_vectored(
+                                    &key_request_accum.consume_keyspace().ranges,
+                                    lsn,
+                                    ctx,
+                                )
+                                .await?;
+
+                            for (img_key, img) in results {
```

Comment:
Hm, another level of indent.
Can we do this more elegantly?
I also don't like the clutter introduced by the `if ...key space full or at end of range ... { ... do vectored get ...; ... process results ...}`.

E.g., did you think about modelling it as iterators / streams?

```
let key_batches = ...;
let img_debatched = key_batches.into_iter().map(|batch| Timeline::get_vectored(batch, ...));
for (img_key, img) in img_debatched  {
  // existing code
}
```

For the record, I don't know how much of a pain this would be with async Rust.

Reply:
I think I can flatten the outer for and while loops by implementing a chunks style iterator for `KeySpace`. I'll defer this to another PR unless you feel strongly about it.